### PR TITLE
FireEyeApiModule and FireEye Email Security

### PR DESCRIPTION
**Pull request by bakatzir**
## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/14274

## Description
Consolidated an APIModule for FE
use it in FE CM
use it in FE EX

## Screenshots
in the referenced issue

## Minimum version of Cortex XSOAR
- [x] 5.5.0

## Does it break backward compatibility?
- [x] No

## Must have
- [x] Tests
- [x] tests for watchlists
- [x] Documentation 
- [x] url (slash must be prefixed by double backslash)
- [x] make sure to check intermittent 401

### CODE REVIEW

File: Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            'Accept': 'application/json',
+        }
+
+    @logger
+    def _generate_token(self) -> str:
+        resp = self._http_request(method='POST', url_suffix='auth/login', resp_type='response')
+        if resp.status_code != 200:
+            raise DemistoException(
+                f'Token request failed with status code {resp.status_code}. message: {str(resp)}')
+        return resp.headers['X-FeApi-Token']
```
Comment (yuvalbenshalom) [inline Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.py:24]: consider adding check here that the key `X-FeApi-Token` really exist and add raise an exception if not.

File: Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        return self._http_request(method='GET', url_suffix='alerts', params=request_params, resp_type='json')
+
+    @logger
+    def get_alert_details_request(self, alert_id: str, timeout: int) -> Dict[str, str]:
+        return self._http_request(method='GET', url_suffix=f'alerts/alert/{alert_id}', resp_type='json',
+                                  timeout=timeout)
+
+    @logger
+    def alert_acknowledge_request(self, uuid: str) -> Dict[str, str]:
+        # data here is redundant, but without it we are getting an error.
```
Comment (yuvalbenshalom) [inline Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.py:37]: it is unclear what data is redundant

File: Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            'start_time': start_time,
+            'end_time': end_time,
+            'limit': limit
+        }
+        if from_:
+            params['from'] = from_
+        if subject:
+            params['subject'] = subject
+        if appliance_id:
+            params['appliance_id'] = appliance_id
```
Comment (yuvalbenshalom) [inline Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.py:78]: in the end there is no receiver?

File: Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.yml [hunk 4]
```diff
@@ -0,0 +1,921 @@
...
+    - contextPath: FireEyeEX.matches
+      description: The Number of matches for the domain name against the incoming emails.
+      type: string
+  isfetch: true
+  runonce: false
+  script: '-'
+  type: python
+  subtype: python3
+  dockerimage: demisto/python3:3.9.5.21272
+fromversion: 5.5.0
```
Comment (yuvalbenshalom) [inline Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.yml:919]: we can probably do it from 6 now

File: Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.py [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+    raw_response = client.fe_client.get_alerts_request(request_params={
+        'start_time': to_fe_datetime_converter(first_fetch),
+        'info_level': info_level
+    })
+    all_alerts = raw_response.get('alert')
+
+    if not all_alerts:
+        demisto.info(f'{INTEGRATION_NAME} no alerts were fetched at: {str(next_run)}')
+        # as no alerts occurred till now, update last_run time accordingly
+        last_run['time'] = to_fe_datetime_converter('now')
```
Comment (yuvalbenshalom) [inline Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.py:471]: if we did not fetch any alerts, we can just use the time from the run before. using time from noe is creating a race new alerts that can be created now.

also, I think you should use next run, but not sure about `last_alert_ids`

File: Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.py [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+    if not incidents:
+        demisto.info(f'{INTEGRATION_NAME} no new alerts were fetched at: {str(next_run)}')
+        # as no alerts occurred till now, update last_run time accordingly
+        last_run['time'] = to_fe_datetime_converter('now')
+        return last_run, []
+
+    # as alerts occurred till now, update last_run time accordingly to the that of latest fetched alert
+    next_run = {
+        'time': alerts[-1].get('occurred'),
+        'last_alert_ids': json.dumps(last_alert_ids)  # save the alert IDs from the last fetch
```
Comment (yuvalbenshalom) [inline Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.py:501]: in wat case `last_alert_ids` is being cleaned?

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
looks good.
see my comments.

File: Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.yml [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+type: python
+subtype: python3
+tags:
+- infra
+- server
+comment: Common FireEye code that will be appended to each FireEye integration when it is deployed.
+scripttarget: 0
+dependson: {}
+timeout: 0s
+dockerimage: demisto/python3:3.8.6.12176
```
[the remark this answers is not in the document]
Comment (bakatzir) [reply Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.yml:15]: it has no sense, and no affect either.

File: Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.py (hunk 3 above)
Comment (bakatzir) [reply Packs/ApiModules/Scripts/FireEyeApiModule/FireEyeApiModule.py:81]: no, not optional via the API.

File: Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.py (hunk 5 above)
Comment (bakatzir) [reply Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.py:471]: the problem is that the `duration` of the search can not be more than 48 hours(api limitation), which is the reason i'm advancing the time even if no alerts were fetched.
agrees with the 2nd point.

File: Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.py (hunk 6 above)
Comment (bakatzir) [reply Packs/FireEyeEX/Integrations/FireEyeEX/FireEyeEX.py:501]: currently it does not, just as we did in [Zimperium](https://github.com/demisto/content/blob/master/Packs/Zimperium/Integrations/Zimperium/Zimperium.py#L610) and in arcsight as well IIRC.

**Comment by ShirleyDenkberg** [comment]
@yuvalbenshalom Doc review completed.

**Review by yuvalbenshalom — APPROVED** [review]

**Pull request merged**